src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -415,6 +426,11 @@ private void ShutdownWrite(Exception? shutdownReason)
 
         public override async ValueTask DisposeAsync()
         {
+            if (_stream == null)
+            {
+                return;
+            }
+
             CanReuse = _stream.CanRead && _stream.CanWrite
```

Comment:
Having to inject a callback into the middle of Dispose smells like Dispose is doing too much work and should be broken up. It's the Complete and _processingTask parts you need to run after, right? Can you see if it makes sense for these parts to be moved to a Stop/DrainAsync API that's called before Dispose?

```
            _originalTransport.Input.Complete();
            _originalTransport.Output.Complete();

            await _processingTask;
```

Reply:
I agree. The problem is layering.

There needs to be a public API that the Quic transport and Kestrel share. There isn't an `IConnectionCompleteDrainFeature` unless we add one. This callback API gets the job done and already exists. There is a bit of overhead because the callback is registered, executed, and then immediately removed each time.

Reply:
Ah, Http3StreamContext.StreamContext is a ConnectionContext, not a QuicStreamContext, so the caller wouldn't have access to any new API unless we added it to the base type. Given this usage pattern that still might be worth considering. Is adding one new virtual method on the base type better than a callback pattern?

Do you anticipate other uses for the IConnectionCompleteFeature?

Reply:
I doubt it.

IMO we can leave adding a new feature, or method, to after .NET 6.